File: algorithms/sort/quickSort.go
package sort

//快速排序充分利用了分而治之的思路

func QSort(arr []int) []int {
	if len(arr) < 2 {
		return arr
	}
	// base line
	pivot := arr[0]
	var left, right []int
	for _, e := range arr[1:] {
		if e <= pivot {
			left = append(left, e)
		} else {
			right = append(right, e)
		}
	}
	return append(QSort(left), append([]int{pivot}, QSort(right)...)...)
}

File: sort_dart/quickSort.dart
List quickSort(List arr) {
  // 快速排序，使用的其实就是分而治之的比较
  if (arr.length < 2) {
    return arr;
  }
  var pivot = arr[0];
  List left = [];
  List right = [];
  for (var i = 1; i < arr.length; i++) {
    if (pivot > arr[i]) {
      left.add(arr[i]);
    } else {
      right.add(arr[i]);
    }
  }

  return [...quickSort(left), pivot, ...quickSort(right)];
}

void main(List<String> args) {
  print(quickSort([2, 1, 9, 7, 6, 5]));
}
